cpp/src/jit/parser.cpp

```diff
@@ -386,53 +324,36 @@ std::string ptx_parser::parse()
   std::string no_comments = remove_comments(ptx);
 
   input_arg_list.clear();
-  // Go directly to the .func mark
-  const size_t length = no_comments.size();
-  size_t start        = no_comments.find(".func") + 5;
-  if (start == length + 5) {
+  auto const _func = std::string(".func");  // Go directly to the .func mark
+  auto f = std::search(no_comments.cbegin(), no_comments.cend(), _func.cbegin(), _func.cend()) +
+           _func.size();
+  if (f >= no_comments.cend()) {
     printf("No function (.func) found in the input ptx code.\n");
     exit(1);
   }
-  size_t stop = start;
-  while (stop < length && no_comments[stop] != '{') { stop++; }
-
-  std::string function_header = std::string(no_comments, start, stop - start);
-
-  stop++;
-  start = stop;
-
-  int bra_count = 0;
-  while (stop < length) {
-    if (no_comments[stop] == '{') bra_count++;
-    if (no_comments[stop] == '}') {
-      if (bra_count == 0) {
-        break;
-      } else {
-        bra_count--;
-      }
-    }
-    stop++;
-  }
 
-  std::vector<std::string> function_body_output =
-    parse_function_body(std::string(no_comments, start, stop - start));
+  auto l = std::find(f, no_comments.cend(), '{');
 
-  std::string function_header_output = parse_function_header(function_header);
-
-  std::string final_output = function_header_output + "\n";
-  final_output += " asm volatile (\"{\");";
-  for (int i = 0; i < function_body_output.size(); i++) {
-    if (function_body_output[i].find("ret;") != std::string::npos) {
-      final_output += "  asm volatile (\"bra RETTGT;\");\n";
-      continue;
+  auto f2 = std::next(l);
+  auto l2 = std::find_if(f2, no_comments.cend(), [brace_count = 0](auto c) mutable {
+    if (c == '{') ++brace_count;
+    if (c == '}') {
+      if (brace_count == 0) return true;  // find matching } to first found {
+      --brace_count;
     }
-    final_output += "  " + function_body_output[i] + "\n";
-  }
-  final_output += " asm volatile (\"RETTGT:}\");";
-
-  final_output += "}";
-
-  return final_output;
+    return false;
+  });
+
+  // DO NOT CHANGE ORDER - parse_function_body must be called before parse_function_header
```

Comment:
The function body needs to be parse first because the function parameter types are inferred from their corresponding load instructions in the function body.

Reply:
I will update the comment to reflect that. Thanks!